02_C# Fundamentals/Task_MyIoC/MyIoC/Container.cs

```diff
@@ -0,0 +1,150 @@
+﻿using MyIoC.Exceptions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace MyIoC
+{
+    public class Container
+    {
+        private readonly IDictionary<Type, Type> _typesDictionary;
+
+        public Container()
+        {
+            _typesDictionary = new Dictionary<Type, Type>();
+
+        }
+
+        public void AddAssembly(Assembly assembly)
+        {
+            var types = assembly.ExportedTypes;
+            foreach (var type in types)
+            {
+                var constructorImportAttribute = type.GetCustomAttribute<ImportConstructorAttribute>();
+                if (constructorImportAttribute != null || HasImportProperties(type))
+                {
+                    _typesDictionary.Add(type, type);
+                }
+
+                var exportAttributes = type.GetCustomAttributes<ExportAttribute>();
+                foreach (var exportAttribute in exportAttributes)
+                {
+                    _typesDictionary.Add(exportAttribute.Contract ?? type, type);
+                }
+            }
+        }
+
+        private bool HasImportProperties(Type type)
+        {
+            var propertiesInfo = GetPropertiesRequiedImport(type);
+            return propertiesInfo.Any();
+        }
+
+        public void AddType(Type type)
+        {
+            _typesDictionary.Add(type, type);
+        }
+
+        public void AddType(Type type, Type baseType)
+        {
+            _typesDictionary.Add(baseType, type);
+        }
+
+        public object CreateInstance(Type type)
+        {
+            if (!_typesDictionary.ContainsKey(type))
+            {
+                throw new IoCException($"IoC container does not have a dependency for {type.FullName}");
+            }
+
+            Type concreteType = _typesDictionary[type];
+            ConstructorInfo constructorInfo = GetConstructor(concreteType);
+            object instance = CreateFromConstructor(concreteType, constructorInfo);
+
+            if (concreteType.GetCustomAttribute<ImportConstructorAttribute>() != null)
+            {
+                return instance;
+            }
```

Comment:
Why don't you do `ResolveProperties` for types marked with `ImportConstructorAttribute`? They still can have `ImportAttributes` on some of their properties, that also need to be imported.

Also, if I'm not wrong, you should only inject parameters in constructor for types marked with `ImportConstructorAttribute`. Otherwise, this attribute does not make much sense.

Reply:
> Why don't you do `ResolveProperties` for types marked with `ImportConstructorAttribute`? They still can have `ImportAttributes` on some of their properties, that also need to be imported.

I think we do not have to expect this because we have a limitation in the task.

Reply:
Alright, considering the situation I don't see a reason to spend even more time on this. If we had more time, I would ask you to ignore this artificial limitation and make your program more reasonable and usable. For now let's finish this